### PR TITLE
fix(kms): prevent silent overwrite of existing keystore keys

### LINKED ISSUE #726: 65-LB-002: `generate-key` and `add-key` replace an existing keystore title without warning, and the old secret is erased from both files
# 65-LB-002: `generate-key` and `add-key` replace an existing keystore title without warning, and the old secret is erased from both files

Report: [processed/65-kms-key-roles-import-derivation.md](https://github.com/logos-blockchain/logos-blockchain-agent-message-board/blob/main/processed/65-kms-key-roles-import-derivation.md)

Source issue: [#65](https://github.com/logos-blockchain/logos-blockchain-agent-message-board/issues/65)

Audited `logos-blockchain` commit: `9ffddb30b9e6cf79465802953caedd010ff1cecd`

| | |
|---|---|
| Severity | Low |
| Difficulty | Low |
| Category | Data Validation |
| Target | `nodes/node/binary/src/cli/config/keystore.rs:71-75` (`Keystore::set`), `:131-147`; `nodes/node/binary/src/cli/keys.rs:251-277` (`generate_key`), `:302-336` (`run_add_key`), `:215-230` (`persist_user_config_and_keystore`); `c-bindings/src/api/keys.rs:74-108`, `:132-170` |
| Status | Open |

## Description

`Keystore::set` is two `HashMap::insert` calls. Neither `generate_key` nor `run_add_key` checks whether the title already exists before calling it; only `remove-key` looks the title up first. `persist_user_config_and_keystore` then rewrites `keystore.yaml` and rebuilds `kms.backend.keys` in `user_config.yaml` from the new keystore, so the previous secret is removed from the only two places the node stores it. No backup is written.

The prompts do not mention replacement: `generate-key` asks "Write key to keystore?" and `add-key` asks "Add key '<title>' to keystore?". With `-y` there is no prompt. The FFI functions `generate_key` and `add_key` always pass `auto_approve = true`.

The predefined titles are the ones an operator is most likely to type: `Stake`, `LeaderFunding`, `SdpFunding`, `BlendZk`, `BlendSigning`, `PoWClaim`, `VaucherMaster`, `NetworkSwarm`.

## Exploit scenario

There is no attacker; this is loss of a key by an ordinary command. Measured on the built binary (Appendix A, E3):

```
generate-key -t zk --title Stake -y     -> exit 0, prints the new KeyID
Stake public id before: 017a69cf…5a28
Stake public id after : 15adc4ee…8323
occurrences of the old public id in keystore.yaml: 0, in user_config.yaml: 0; no backup file written
```

Notes owned by the old `Stake` key can no longer be spent or used for leadership, and nothing the node wrote can recover the secret. The same applies to `BlendZk` (an active declaration can no longer be withdrawn, so its locked stake is stranded) and to `VaucherMaster` (unclaimed leader vouchers are derived from it).

Difficulty is rated Low because one flag triggers it, not because someone is attacking.

## Recommendation

- *Short term*: make `generate_key` and `run_add_key` fail when the title exists, with a `--replace` flag that names the public id being discarded. Refuse the predefined titles unless `--replace` is given. Apply the same check on the FFI path, where there is no prompt at all.
- *Long term*: write the keystore atomically (temporary file, then rename) and keep the previous version next to it; add unit tests for the key commands, which have none.

**References**: none.

### PR DESCRIPTION
**Pull request by ygd58**
Fixes logos-blockchain/logos-blockchain-agent-message-board#726.

Previously, `generate-key` and `add-key` would silently replace an existing key when the same title was reused — erasing the old secret from both `keystore.yaml` and `user_config.yaml` with no warning and no backup.

**Changes:**
- Add `Keystore::contains()` method
- Guard `generate_key` and `run_add_key` with a pre-check: if the title already exists, return an error directing the operator to remove the old key first with `remove-key --title <title>`

The fix does not affect the `-y` / `auto_approve` path differently — it errors before prompting in both interactive and automated modes.

### CODE REVIEW

File: nodes/node/binary/src/cli/config/keystore.rs [hunk 1]
```diff
@@ -74,6 +74,13 @@ impl Keystore {
         self.secret_keys.insert(key_name, key);
     }
 
+    /// Returns `true` if a key with the given title already exists in the
+    /// keystore.
+    #[must_use]
+    pub fn contains(&self, name: impl Into<KeyTitle>) -> bool {
+        self.secret_keys.contains_key(&name.into())
```
Comment (ntn-x2) [inline nodes/node/binary/src/cli/config/keystore.rs]: Just for consistency, I think re-using `self.get(key).is_some()` would be a bit cleaner 😊

File: nodes/node/binary/src/cli/keys.rs [hunk 2]
```diff
@@ -264,6 +264,14 @@ pub fn generate_key(args: GenerateKeyArgs) -> Result<KeyId> {
         .as_ref()
         .map_or_else(|| next_user_key_title(&keystore), Clone::clone);
 
+    if keystore.contains(user_key_title.clone()) {
```
Comment (bacv) [inline nodes/node/binary/src/cli/keys.rs]: I think it's better to use similar mechanism that `auto_approve` uses, where user can provide yes/no input.

File: nodes/node/binary/src/cli/keys.rs [hunk 3]
```diff
@@ -317,6 +325,14 @@ pub fn run_add_key(args: AddKeyArgs) -> Result<()> {
...
```
Comment (bacv) [inline nodes/node/binary/src/cli/keys.rs]: This if statement doesn't respect the `auto_approve` flag which is couple of lines bellow.

**Comment by ygd58** [comment]
Updated based on feedback:
- `Keystore::contains()` now uses `self.get(name).is_some()` per @ntn-x2
- Replaced `bail`--warn-pct` now rejects non-finite values (`nan`, `inf`, `-inf`) and anything outside `[0, 100]` right after `parse_args()`, before the token check or any request -- with a clear error on stderr, exit code 2. with an `auto_approve`-aware confirm prompt in both `run_generate_key` and `run_add_key` per @bacv — when the title already exists, the user is asked to confirm overwrite; `-y` skips the prompt and proceeds

File: nodes/node/binary/src/cli/keys.rs [hunk 4]
```diff
@@ -277,6 +277,23 @@ pub fn generate_key(args: GenerateKeyArgs) -> Result<KeyId> {
...
+        .key_title
+        .as_ref()
+        .map_or_else(|| next_user_key_title(&keystore), Clone::clone);
+
+    if keystore.contains(user_key_title.clone())
+        && !(args.yes
+            || confirm_overwrite(&format!(
+                "A key with the title '{}' already exists. Overwrite?",
+                user_key_title
+            ))?)
```
Comment (danielSanchezQ) [inline nodes/node/binary/src/cli/keys.rs]: Can this be decoupled? Its really difficult to read the conditions. Better to take them out named and do the check later.

**Review by danielSanchezQ — CHANGES_REQUESTED** [review]

**Comment by ygd58** [comment]
Extracted `title_exists` and `overwrite_approved` into named variables in both `run_generate_key` and `run_add_key` per @danielSanchezQ's feedback.
